hibernate-core/src/main/java/org/hibernate/persister/entity/AbstractEntityPersister.java

```diff
@@ -5167,34 +5168,36 @@ public void setPropertyValues(Object object, Object[] values) {
 		else {
 			if ( hasSubclasses() ) {
 				visitAttributeMappings(
-						attribute -> {
-							final int stateArrayPosition = ( (StateArrayContributorMapping) attribute ).getStateArrayPosition();
-							final Object value = values[stateArrayPosition];
-							if ( value != UNFETCHED_PROPERTY ) {
-								final Setter setter = attribute.getPropertyAccess().getSetter();
-								setter.set( object, value, getFactory() );
-							}
-						}
+						attribute -> setPropertyValueInternally( attribute, object, values )
 				);
 			}
 			else {
 				visitFetchables(
-						fetchable -> {
-							final AttributeMapping attribute = (AttributeMapping) fetchable;
-							final int stateArrayPosition = ( (StateArrayContributorMapping) attribute ).getStateArrayPosition();
-							final Object value = values[stateArrayPosition];
-							if ( value != UNFETCHED_PROPERTY ) {
-								final Setter setter = attribute.getPropertyAccess().getSetter();
-								setter.set( object, value, getFactory() );
-							}
-
-						},
+						fetchable -> setPropertyValueInternally( (AttributeMapping) fetchable, object, values ),
 						null
 				);
 			}
 		}
 	}
 
+	private void setPropertyValueInternally(AttributeMapping attribute, Object object, Object[] values) {
```

Comment:
intentionally avoid naming of 'setPropertyValue' for it will end up with method overloading with the public method just below.